### ISSUE TITLE
osd prepare pod fails with ceph version v14.2.3-20190904

### ISSUE THREAD

**Issue by sp98** [rook/rook#3795]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
OSD prepare pod fails to start 

**Expected behavior:**
OSD prepare pod should start successfully. 

**How to reproduce it (minimal and precise):**

1. apply common.yaml
2. apply operator.yaml.
3. apply cluster-on-pvc.yaml with ceph image - ceph/ceph:v14.2.3-20190904
4. Observer that the OSD prepare job fails

Note: issue is observe with ceph  version - v14.2.3-20190904.  Version 14.2.2 is working fine. 

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

 To get logs, use `kubectl -n <namespace> logs <pod name>`
When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
Read [Github documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Storage backend version (e.g. for ceph do `ceph -v`): :v14.2.3-20190904
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift): openshift
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):

**Comment by sp98** [comment]
Possible root cause is root is no longer reading the output of the `ceph lvm prepare` command. 
The output of this command is used to fetch the logical volume path created by ceph-volume. OSD prepare pod is failing because the output is empty!

**Comment by BlaineEXE** [comment]
At first glance, this feels like an issue in Ceph to me.

**Comment by BlaineEXE** [comment]
Related to this? https://tracker.ceph.com/issues/41660

**Comment by sp98** [comment]
ceph-volume lvm prepare logs for 14.2.3
```[root@rook-ceph-osd-prepare-set1-0-data-p545h-4drcb /]# ceph-volume  lvm prepare --data /mnt/set1-0-data-p545h > stdout 2> stderr
[root@rook-ceph-osd-prepare-set1-0-data-p545h-4drcb /]# cat stdout
[root@rook-ceph-osd-prepare-set1-0-data-p545h-4drcb /]# cat stderr
 stderr: Unknown device, --name=, --path=, or absolute path in /dev/ or /sys expected.
Running command: /bin/ceph-authtool --gen-print-key
Running command: /bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring -i - osd new f5afd38a-4a53-4cf9-8eb6-829e9a49dfb7
Running command: /usr/sbin/vgcreate -s 1G --force --yes ceph-d05955e4-317f-4423-9c61-425192f2327c /mnt/set1-0-data-p545h
 stdout: Physical volume "/mnt/set1-0-data-p545h" successfully created.
 stdout: Volume group "ceph-d05955e4-317f-4423-9c61-425192f2327c" successfully created
Running command: /usr/sbin/lvcreate --yes -l 100%FREE -n osd-block-f5afd38a-4a53-4cf9-8eb6-829e9a49dfb7 ceph-d05955e4-317f-4423-9c61-425192f2327c
 stderr: WARNING: Not using device /dev/xvdcg for PV PyYfmf-nzmg-gXQG-n4AI-25Dr-i2qh-59caJC.
  WARNING: PV PyYfmf-nzmg-gXQG-n4AI-25Dr-i2qh-59caJC prefers device /dev/loop0 because device was seen first.
 stdout: Logical volume "osd-block-f5afd38a-4a53-4cf9-8eb6-829e9a49dfb7" created.
Running command: /bin/ceph-authtool --gen-print-key
Running command: /bin/mount -t tmpfs tmpfs /var/lib/ceph/osd/ceph-0
Running command: /usr/sbin/restorecon /var/lib/ceph/osd/ceph-0
Running command: /bin/chown -h ceph:ceph /dev/ceph-d05955e4-317f-4423-9c61-425192f2327c/osd-block-f5afd38a-4a53-4cf9-8eb6-829e9a49dfb7
Running command: /bin/chown -R ceph:ceph /dev/mapper/ceph--d05955e4--317f--4423--9c61--425192f2327c-osd--block--f5afd38a--4a53--4cf9--8eb6--829e9a49dfb7
Running command: /bin/ln -s /dev/ceph-d05955e4-317f-4423-9c61-425192f2327c/osd-block-f5afd38a-4a53-4cf9-8eb6-829e9a49dfb7 /var/lib/ceph/osd/ceph-0/block
Running command: /bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring mon getmap -o /var/lib/ceph/osd/ceph-0/activate.monmap
 stderr: got monmap epoch 3
Running command: /bin/ceph-authtool /var/lib/ceph/osd/ceph-0/keyring --create-keyring --name osd.0 --add-key AQBp2HVdsT3MBhAACIBATe7/AsiQnkDJAFuhcg==
 stdout: creating /var/lib/ceph/osd/ceph-0/keyring
added entity osd.0 auth(key=AQBp2HVdsT3MBhAACIBATe7/AsiQnkDJAFuhcg==)
Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0/keyring
Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0/
Running command: /bin/ceph-osd --cluster ceph --osd-objectstore bluestore --mkfs -i 0 --monmap /var/lib/ceph/osd/ceph-0/activate.monmap --keyfile - --osd-data /var/lib/ceph/osd/ceph-0/ --osd-uuid f5afd38a-4a53-4cf9-8eb6-829e9a49dfb7 --setuser ceph --setgroup ceph
--> ceph-volume lvm prepare successful for: /mnt/set1-0-data-p545h
```

ceph-volume lvm prepare logs for 14.2.2
```
[root@rook-ceph-osd-prepare-set1-0-data-4jdrp-vql4t /]# ceph-volume  lvm prepare --data /mnt/set1-0-data-4jdrp > stdout 2> stderr
[root@rook-ceph-osd-prepare-set1-0-data-4jdrp-vql4t /]# cat stdout
 stderr: Unknown device, --name=, --path=, or absolute path in /dev/ or /sys expected.
Running command: /bin/ceph-authtool --gen-print-key
Running command: /bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring -i - osd new aa4e5390-f568-4d7a-a48a-39ecdb828f9f
Running command: /usr/sbin/vgcreate -s 1G --force --yes ceph-fecf440c-e0ab-43c5-b304-35613b3f20ef /mnt/set1-0-data-4jdrp
 stdout: Physical volume "/mnt/set1-0-data-4jdrp" successfully created.
 stdout: Volume group "ceph-fecf440c-e0ab-43c5-b304-35613b3f20ef" successfully created
Running command: /usr/sbin/lvcreate --yes -l 100%FREE -n osd-block-aa4e5390-f568-4d7a-a48a-39ecdb828f9f ceph-fecf440c-e0ab-43c5-b304-35613b3f20ef
 stderr: WARNING: Not using device /dev/xvdcp for PV eiawkZ-ec1P-1FxS-ReXI-Rm6R-60f0-NNweX4.
  WARNING: PV eiawkZ-ec1P-1FxS-ReXI-Rm6R-60f0-NNweX4 prefers device /dev/loop0 because device was seen first.
 stdout: Logical volume "osd-block-aa4e5390-f568-4d7a-a48a-39ecdb828f9f" created.
Running command: /bin/ceph-authtool --gen-print-key
Running command: /bin/mount -t tmpfs tmpfs /var/lib/ceph/osd/ceph-0
Running command: /usr/sbin/restorecon /var/lib/ceph/osd/ceph-0
Running command: /bin/chown -h ceph:ceph /dev/ceph-fecf440c-e0ab-43c5-b304-35613b3f20ef/osd-block-aa4e5390-f568-4d7a-a48a-39ecdb828f9f
Running command: /bin/chown -R ceph:ceph /dev/mapper/ceph--fecf440c--e0ab--43c5--b304--35613b3f20ef-osd--block--aa4e5390--f568--4d7a--a48a--39ecdb828f9f
Running command: /bin/ln -s /dev/ceph-fecf440c-e0ab-43c5-b304-35613b3f20ef/osd-block-aa4e5390-f568-4d7a-a48a-39ecdb828f9f /var/lib/ceph/osd/ceph-0/block
Running command: /bin/ceph --cluster ceph --name client.bootstrap-osd --keyring /var/lib/ceph/bootstrap-osd/ceph.keyring mon getmap -o /var/lib/ceph/osd/ceph-0/activate.monmap
 stderr: got monmap epoch 3
Running command: /bin/ceph-authtool /var/lib/ceph/osd/ceph-0/keyring --create-keyring --name osd.0 --add-key AQDF33VdNgUQOxAAAHeNQxIRr2mrMQc9P3zLqQ==
 stdout: creating /var/lib/ceph/osd/ceph-0/keyring
added entity osd.0 auth(key=AQDF33VdNgUQOxAAAHeNQxIRr2mrMQc9P3zLqQ==)
Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0/keyring
Running command: /bin/chown -R ceph:ceph /var/lib/ceph/osd/ceph-0/
Running command: /bin/ceph-osd --cluster ceph --osd-objectstore bluestore --mkfs -i 0 --monmap /var/lib/ceph/osd/ceph-0/activate.monmap --keyfile - --osd-data /var/lib/ceph/osd/ceph-0/ --osd-uuid aa4e5390-f568-4d7a-a48a-39ecdb828f9f --setuser ceph --setgroup ceph
--> ceph-volume lvm prepare successful for: /mnt/set1-0-data-4jdrp
[root@rook-ceph-osd-prepare-set1-0-data-4jdrp-vql4t /]# cat stderr
[root@rook-ceph-osd-prepare-set1-0-data-4jdrp-vql4t /]# 

```

**Comment by jan--f** [comment]
Correct, 14.2.3 introduced a change, where ceph-volume stops printing status/log messages to stdout and prints them to stderr instead.

**Comment by sp98** [comment]
@jan--f Thanks for confirming this . By "change" do you mean this is a new  and expected behavior, or an issue that will be addressed soon?

**Comment by jan--f** [comment]
> @jan--f Thanks for confirming this . By "change" do you mean this is a new and expected behavior, or an issue that will be addressed soon?

This new behaviour is intended and expected.

**Comment by travisn** [comment]
@sp98 If the output is going to stderr, this should be a quick fix in rook to call `ExecuteCommandWithCombinedOutput()` instead of `ExecuteCommandWithOutput()` when calling ceph-volume

**Comment by BlaineEXE** [comment]
Question: why are we parsing the output of this command in Rook as we do? This has exposed the fragility of parsing the CLI output of commands.

**Comment by travisn** [comment]
@BlaineEXE This was the only way in the short term to get results from the ceph-volume command. What we need from `ceph-volume` is that results in json format so we can parse it reliably. @rohan47 Did you already discuss with alfredo what this would take and/or open an issue?

**Comment by BlaineEXE** [comment]
@jan--f is the new maintainer of `ceph-volume`. Does Alfredo still work the project at all?

**Comment by travisn** [comment]
Right, that's an old habit to ping alfredo for c-v issues. :) @rohan47 can you sync up with @jan--f?

**Comment by alfredodeza** [comment]
I certainly still do ceph-volume work :) Happy to assist here if needed, but it is fine also to talk directly to Jan.

**Comment by alfredodeza** [comment]
Ceph has released 14.2.4 which includes the fixes for stderr/stdout encoding, which may be related to some of the problems in this issue even though PR #3811 combines the output.

The relevant tracker issue is at https://tracker.ceph.com/issues/41660
